portlayer/storage/vsphere/store.go

```diff
@@ -254,8 +256,24 @@ func (v *ImageStore) WriteImage(ctx context.Context, parent *portlayer.Image, ID
 		}
 		defer vmdisk.Unmount()
 
-		// Untar the archive
-		cerr = archive.Untar(r, dir, &archive.TarOptions{})
+		// Decompress the archive into memory (this would happen anyway within the call to Untar)
+		decompressedTar, cerr := archive.DecompressStream(r)
+		if cerr != nil {
+			return nil, cerr
+		}
+
+		// Get the SHA256 sum of the decompressed archive (layer diffID)
+		h := sha256.New()
+		if _, cerr := io.Copy(h, decompressedTar); cerr != nil {
```

Comment:
This will copy all of the data presented by the reader (`decompressedTar`) to h.  You will get your sha that way, but the untar below will likely get nothing.  If you use a TeeReader (https://golang.org/pkg/io/#TeeReader), you can avoid the `Copy` of the whole of the data and get the sha inline.  Pretty sure we already do exactly that in `store_cache.go`.  Either move that bit of code (TeeReader for the xfer sha sum) here or do this there.  In either case, `io.Copy` may not be what you want.